### DISCUSSION TITLE
file system download from a remote server or load from a custom location when compiling

### DISCUSSION THREAD

**Discussion by urechi** [arendst/Tasmota#16547] [Ideas]
My use case:
I manage several tasmota zigbee hubs for friends and family through MQTT. I don't have access to the web UI to be able to upload files to the filesystem and for security reasons the UIs are not exposed to the internet. But I have a need to upload berry and zigbee (for rare attribute mapping) files to those hubs' filesystems. If this is not possible to do remotely in a safe way then I propose that files from a custom directory from the repo, for example `tasmota/fs_uploads/` are copied automatically to fs when compiling your own version of tasmota. Any thoughts?

**Comment by urechi** [comment]
Or a command similar to `WebGetConfig`, to allow the download of files from a remote server to the filesystem?

**Reply by barbudor** [reply]
There's a option to embed files in the filesystem when compiling. I'm not much aware of the details but @Jason2866 can help

A WebDownload to file system command could make sense

I think there are a way to transport files over mqtt but I have not explored that and can't find in the docs. Would need to dig a bit.

But if you are using ESP32 based zigbee, the most efficient would probably to make a small Berry script to download from a web storage of yours upon request.

I could have a look to implementing WebDownload command next week

**Reply by sfromis** [reply]
What I'm doing is to simulate having a "WebDownload" by running a tiny Berry script to fetch a file from some URL. I could have that Berry code stored in the file system, but instead I generate a command including the script code, meaning that I can simply paste that command into the Berry console, or via the `Br` command run the code from the Tasmota console, or via MQTT, HTTP or serial. Here's the code with whitespace added for readability:
```
def urlfetch(url,file)
  if file==nil
    import string
    file=string.split(url,'/').pop()
  end
  var wc=webclient()
  wc.begin(url)
  var st=wc.GET()
  if st!=200 raise 'connection_error','status: '+str(st) end
  st='Fetched '+str(wc.write_file(file))
  print(url,st)
  wc.close()
  return st
end
```
The command becomes that code in one line, and something like this added at the end:
`urlfetch('https://github.com/arendst/Tasmota/blob/development/tasmota/zigbee/giex_water.zb')`
The `return st` at the end is to have a nice response included if executed via the `Br` command.

(I already mentioned the convenience of having this as a built-in to @s-hadinger some time ago)

**Reply by urechi** [reply]
@sfromis I guess I will need `;` after every line when running it as a single line command with `br` ?

**Reply by urechi** [reply]
```
11:55:49.784 CMD: br def urlfetch(url,file); if file==nil; import string; file=string.split(url,'/').pop(); end; var wc=webclient(); wc.begin(url); var st=wc.GET(); if st!=200 raise 'connection_error','status: '+str(st) end; st='Fetched '+str(wc.write_file(file)); print(url,st); wc.close(); return st; end;
11:55:49.786 SRC: WebConsole from 192.168.1.44
11:55:49.787 CMD: Grp 0, Cmd 'BR', Idx 1, Len 286, Pld -99, Data 'def urlfetch(url,file); if file==nil; import string; file=string.split(url,'/').pop(); end; var wc=webclient(); wc.begin(url); var st=wc.GET(); if st!=200 raise 'connection_error','status: '+str(st) end; st='Fetched '+str(wc.write_file(file)); print(url,st); wc.close(); return st; end;'
11:55:49.806 MQT: stat/SmartHub-D3F00C/RESULT = {"Br":"nil"}
```

**Reply by urechi** [reply]
```
12:04:57.130 CMD: br def urlfetch(url,file); if file==nil; import string; file=string.split(url,'/').pop(); end; var wc=webclient(); wc.begin(url); var st=wc.GET(); if st!=200 raise 'connection_error','status: '+str(st) end; st='Fetched '+str(wc.write_file(file)); print(url,st); wc.close(); return st; end; urlfetch('https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/zigbee/giex_water.zb')
12:04:57.132 SRC: WebConsole from 192.168.1.44
12:04:57.134 CMD: Grp 0, Cmd 'BR', Idx 1, Len 389, Pld -99, Data 'def urlfetch(url,file); if file==nil; import string; file=string.split(url,'/').pop(); end; var wc=webclient(); wc.begin(url); var st=wc.GET(); if st!=200 raise 'connection_error','status: '+str(st) end; st='Fetched '+str(wc.write_file(file)); print(url,st); wc.close(); return st; end; urlfetch('https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/zigbee/giex_water.zb')'
12:04:57.837 HTP: Connected in 689 ms, stack low mark 896
12:04:57.916 https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/zigbee/giex_water.zb Fetched 949
12:04:57.927 MQT: stat/SmartHub-D3F00C/RESULT = {"Br":"nil"}
```

**Reply by urechi** [reply]
Thanks @sfromis that worked, just a small change:
`urlfetch('https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/zigbee/giex_water.zb')`

**Reply by s-hadinger** [reply]
> @sfromis I guess I will need `;` after every line when running it as a single line command with `br` ?

No you don't need ';'. Berry doesn't care about newlines except for comments. The Berry grammar is almost non ambiguous and doesn't require to separate statements with ';' (except in very rare case that I never saw in real life)

**Reply by urechi** [reply]
Added this [PR](https://github.com/tasmota/docs/pull/1047).

**Reply by s-hadinger** [reply]
Thanks. I think I will add this command after all

**Comment by Jason2866** [comment]
There is an example how files can be included in the filesystem in `platformio_tasmota_cenv_sample.ini` when compiling your own version / env.

```
[env:tasmota32s3-file]
extends                     = env:tasmota32_base
board                       = esp32s3
build_flags                 = ${env:tasmota32_base.build_flags} -D FIRMWARE_TASMOTA32
; example for custom file upload in Tasmota Filesystem
custom_files_upload         = ${env:tasmota32_base.custom_files_upload}
                               tasmota/berry/modules/Partition_wizard.tapp
                               https://github.com/tasmota/autoconf/raw/main/esp32s3/DevKitC-1.autoconf
```
This env is [used](https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env32.ini#L123-L136)

**Comment by urechi** [comment]
Great people, thanks all for your help. If you are ever in Manchester, UK, I'll take you out for some drinks 🍺
@Jason2866 your method is exactly what I was looking for.
@barbudor `WebDownload` would be a nice to have command but as there are workarounds not sure if it's worth the effort.
@sfromis ~~your script would still require access to the web UI initially to be uploaded unless you do it first using @Jason2866 method.~~

**Reply by sfromis** [reply]
No, as I explained, my strategy does **not** require access to the web UI, just any way of issuing a command. Similar to having a `WebDownload`, just a longer command. I clearly explained that I *could* have chosen to placed it in the file system, but as I'm not doing that, I simply generate a long command line for a fresh Tasmota install without any preparation, like installing a script.

**Reply by urechi** [reply]
I see now, apologies, I had no idea that there is a `br` command that you can use to run the code. I've checked the [commands](https://tasmota.github.io/docs/Commands/) page and isn't listed there, maybe worth adding that? In this case, yes your approach is also suitable for this use case. Thanks a lot, I've learned something new and useful, so that's great.

**Comment by urechi** [comment]
Added `moes.zb` file to `tasmota/zigbee/` folder
```
#Z2Tv1
# MOES
# 
# Moes Thermostat
:TS0601, _TZE200_aoclfnxz
EF00/0210=0201/FFF1,mul:100         # temp target in 1/100 °C
EF00/0210,MoesHeatingSetpoint       # attribute used to write value
EF00/0218=0201/0000,mul:10          # Local Temperature in 1/100 °C
EF00/0424,MoesValve
```

Edited my `platformio_tasmota_cenv.ini` file:
```
; *** Tasmota version for EWELINK ZB-GW03 Smart Hub
[env:tasmota32-zb-gw03]
extends                 = env:tasmota32
board_build.f_cpu       = 240000000L
board_build.f_flash     = 40000000L
build_flags             = ${env:tasmota32_base.build_flags} -DEWELINK_ZB_GW03
custom_files_upload     = ${env:tasmota32_base.custom_files_upload}
                          tasmota/zigbee/moes.zb
```

`platformio run -e tasmota32-zb-gw03`
```
Building .pio/build/tasmota32-zb-gw03/firmware.bin
esptool.py v3.3.1
Creating esp32 image...
Merged 25 ELF sections
Successfully created esp32 image.
esp32_create_combined_bin([".pio/build/tasmota32-zb-gw03/firmware.bin"], [".pio/build/tasmota32-zb-gw03/firmware.elf"])
Read partitions from  partitions/esp32_partition_app2880k_fs320k.csv
# Name,     Type,   SubType,   Offset,    Size,   Flags
nvs         data    nvs        0x9000     0x5000
otadata     data    ota        0xe000     0x2000
safeboot    app     factory    0x10000    0xD0000
app0        app     ota_0      0xE0000    0x2D0000
spiffs      data    spiffs     0x3B0000   0x50000
Creating filesystem with content:
/moes.zb
```

Pushed the new firmware through OTA and did a `reset 6` on the hubs but the `moes.zb` is not present in the filesystem.

Am I doing something wrong?

**Reply by Jason2866** [reply]
Yes, you made a false assumption. The filesystem is not uploaded when doing an OTA. The filesystem is only written when flashing the factory firmware via Platformio.

**Reply by sfromis** [reply]
This is a reason why I see this method as a last resort, not useful for device maintenance, but only for creating initial install images

**Reply by urechi** [reply]
Thanks @Jason2866 , I'll give your method a go this afternoon @sfromis. Also @barbudor `WebDownload` makes more sense now.
